test/decompiler/reference/engine/util/smush-control-h_REF.gc

```diff
@@ -0,0 +1,151 @@
+;;-*-Lisp-*-
+(in-package goal)
+
+;; definition of type smush-control
+(deftype smush-control (structure)
+  ((start-time  uint64  :offset-assert 0)
+   (period      float   :offset-assert 8)
+   (duration    float   :offset-assert 12)
+   (amp         float   :offset-assert 16)
+   (damp-amp    float   :offset-assert 20)
+   (damp-period float   :offset-assert 24)
+   (ticks       float   :offset-assert 28)
+   )
+  :method-count-assert 15
+  :size-assert         #x20
+  :flag-assert         #xf00000020
+  (:methods
+    (set-zero! (_type_) _type_ 9)
+    (update! (_type_) float 10)
+    (get-no-update (_type_) float 11)
+    (activate! (_type_ float int int float float float) _type_ 12)
+    (nonzero-amplitude? (_type_) symbol 13)
+    (die-on-next-update! (_type_) _type_ 14)
+    )
+  )
+
+;; definition for method 3 of type smush-control
+(defmethod inspect smush-control ((obj smush-control))
+  (format #t "[~8x] ~A~%" obj 'smush-control)
+  (format #t "~Tstart-time: ~D~%" (-> obj start-time))
+  (format #t "~Tperiod: ~f~%" (-> obj period))
+  (format #t "~Tduration: ~f~%" (-> obj duration))
+  (format #t "~Tamp: ~f~%" (-> obj amp))
+  (format #t "~Tdamp-amp: ~f~%" (-> obj damp-amp))
+  (format #t "~Tdamp-period: ~f~%" (-> obj damp-period))
+  (format #t "~Tticks: ~f~%" (-> obj ticks))
+  obj
+  )
+
+;; definition for method 13 of type smush-control
+(defmethod nonzero-amplitude? smush-control ((obj smush-control))
+  (!= (-> obj amp) 0.0)
+  )
+
+;; definition for method 9 of type smush-control
+(defmethod set-zero! smush-control ((obj smush-control))
+  (set! (-> obj period) 0.0)
+  (set! (-> obj duration) 0.0)
+  (set! (-> obj amp) 0.0)
+  (set! (-> obj damp-amp) 0.0)
+  (set! (-> obj damp-period) 0.0)
+  (set! (-> obj ticks) 0.0)
+  obj
+  )
+
+;; definition for method 10 of type smush-control
+(defmethod update! smush-control ((obj smush-control))
+  (cond
+   ((!= (-> obj amp) 0.0)
+    (let*
+     ((f30-0
+       (the float (- (-> *display* base-frame-counter) (-> obj start-time)))
+       )
+      (f0-2 (-> obj period))
+      (f28-0 (- f30-0 (* (the float (the int (/ f30-0 f0-2))) f0-2)))
+      )
+     (when (>= (- f30-0 (-> obj ticks)) (-> obj period))
+      (set! (-> obj amp) (* (-> obj amp) (-> obj damp-amp)))
+      (set! (-> obj period) (* (-> obj period) (-> obj damp-period)))
+      (set! (-> obj ticks) f30-0)
+      (if (< (-> obj damp-period) 0.0)
+       ((method-of-type smush-control set-zero!) obj)
```

Comment:
This is correct, but the decompiler should simplify it more. Issue here: https://github.com/water111/jak-project/issues/536